cmd/src/dump.rs

```diff
@@ -29,58 +39,116 @@ pub fn check_and_dump_raft_db<E: RaftEngine>(raftdb_path: &str, engine: &E, env:
-    let start_key = keys::REGION_RAFT_MIN_KEY;
-    let end_key = keys::REGION_RAFT_MAX_KEY;
-    let mut log_batch = engine.log_batch(0);
-    let mut count_entry = 0;
-    let mut count_size = 0;
+    let count_size = Arc::new(AtomicUsize::new(0));
     let mut count_region = 0;
-    let consume_time = std::time::Instant::now();
-    origin_engine
-        .scan_cf(CF_DEFAULT, start_key, end_key, false, |key, value| {
-            let res = keys::decode_raft_key(key);
-            match res {
-                Err(_) => Ok(true),
-                Ok((region_id, suffix)) => {
-                    match suffix {
-                        keys::RAFT_LOG_SUFFIX => {
-                            let mut entry = Entry::default();
-                            entry.merge_from_bytes(&value)?;
-                            log_batch.append(region_id, vec![entry]).unwrap();
-                            count_entry += 1;
-                        }
-                        keys::RAFT_STATE_SUFFIX => {
-                            let mut state = RaftLocalState::default();
-                            state.merge_from_bytes(&value)?;
-                            log_batch.put_raft_state(region_id, &state).unwrap();
-                            count_region += 1;
-                        }
-                        // There is only 2 types of keys in raft.
-                        _ => unreachable!(),
-                    }
-                    if count_entry % 1024 == 0 {
-                        let size = engine.consume(&mut log_batch, true).unwrap();
-                        count_size += size;
-                    }
-                    Ok(true)
-                }
+    let mut threads = vec![];
+    let (tx, rx) = unbounded();
+    for _ in 0..thread_num {
+        let raft_engine = engine.clone();
+        let raftdb_engine = origin_engine.clone();
+        let count_size = count_size.clone();
+        let rx = rx.clone();
+        let t = std::thread::spawn(move || {
+            run_worker(rx, raftdb_engine, raft_engine, count_size);
+        });
+        threads.push(t);
+    }
+    info!("Start to scan raft log from raftdb and dump into raft engine");
+    let consumed_time = std::time::Instant::now();
+    // Seek all region id from raftdb and send them to workers.
+    let mut it = origin_engine.iterator().unwrap();
+    let mut valid = it.seek(SeekKey::Key(keys::REGION_RAFT_MIN_KEY)).unwrap();
+    while valid {
+        match keys::decode_raft_key(it.key()) {
+            Err(e) => {
+                error!("Error happened when decoding raft key: {}", e);
+                break;
             }
-        })
-        .unwrap();
-    let size = engine.consume(&mut log_batch, true).unwrap();
+            Ok((id, _)) => {
+                tx.send(id).unwrap();
+                count_region += 1;
+                let next_key = keys::raft_log_prefix(id + 1);
+                valid = it.seek(SeekKey::Key(&next_key)).unwrap();
```

Comment:
nit: better error message on iterator error?

Reply:
When user start failed, he/she should check the log and can position problem.